Mark classes / code as not threadsafe

Hi,

We use spotbugs in a largeish project and there are a number of false positives where spotbugs reports problems with thread (AT_NONATOMIC_64BIT_PRIMITIVE, AT_STALE_THREAD_WRITE_OF_PRIMITIVE, AT_NONATOMIC_OPERATIONS_ON_SHARED_VARIABLE). 

The problem is that the places where this is reported have been explicitly documented as not threadsafe. So problems may exist but it does not matter because using this code from multiple threads is a problem in general.

Is there a way to mark such classes explicitly as not threadsafe so that spotbugs does not try to find multithreading bugs in code that is not intended to be used from multiple threads?

Thanks @JuditKnoll we are aware we can suppress the warnings, however they are so numerous that we would probably just turn these checks off entirely.

Here is an example false positive:

```
import javax.annotation.concurrent.NotThreadSafe;

@NotThreadSafe
public class CountingRunner implements Runnable {
    private int counter;

    public int getCount() {
        return counter;
    }

    @Override
    public void run() {
        counter++;
    }
}
```

fails:
```
[ERROR] Medium: Operation on the "counter" shared variable in "CountingRunner" class is not atomic [CountingRunner] At CountingRunner.java:[line 15] AT_NONATOMIC_OPERATIONS_ON_SHARED_VARIABLE
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,9 @@ Currently the versioning policy of this project follows [Semantic Versioning v2.
 - Do not report `US_USELESS_SUPPRESSION_ON_METHOD` on generated methods ([#3350](https://github.com/spotbugs/spotbugs/issues/3350))
 - Rewrite some member in `ResourceValueFrame.java` to Enum ([#2061](https://github.com/spotbugs/spotbugs/issues/2061))
 
+### Added
+- Multi-threaded code checks can be skipped with `@NotThreadSafe` ([#3390](https://github.com/spotbugs/spotbugs/issues/3390))
+
 ## 4.9.3 - 2025-03-14
 ### Added
 

--- a/spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/Issue3390Test.java
+++ b/spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/Issue3390Test.java
@@ -1,0 +1,12 @@
+package edu.umd.cs.findbugs.detect;
+
+import edu.umd.cs.findbugs.AbstractIntegrationTest;
+import org.junit.jupiter.api.Test;
+
+class Issue3390Test extends AbstractIntegrationTest {
+    @Test
+    void testNotThreadSafe() {
+        performAnalysis("ghIssues/Issue3390NotThreadSafe.class");
+        assertBugInClassCount("AT_NONATOMIC_OPERATIONS_ON_SHARED_VARIABLE", "ghIssues.Issue3390NotThreadSafe", 0);
+    }
+}

--- a/spotbugs/src/main/java/edu/umd/cs/findbugs/detect/SharedVariableAtomicityDetector.java
+++ b/spotbugs/src/main/java/edu/umd/cs/findbugs/detect/SharedVariableAtomicityDetector.java
@@ -87,7 +87,8 @@ public class SharedVariableAtomicityDetector extends OpcodeStackDetector {
 
     @Override
     public void visitClassContext(ClassContext classContext) {
-        if (MultiThreadedCodeIdentifierUtils.isPartOfMultiThreadedCode(classContext)) {
+        if (MultiThreadedCodeIdentifierUtils.isPartOfMultiThreadedCode(classContext)
+                && !MultiThreadedCodeIdentifierUtils.isNotThreadSafe(classContext)) {
             currentMethod = null;
             currentCFG = null;
             currentLockDataFlow = null;

--- a/spotbugs/src/main/java/edu/umd/cs/findbugs/util/MultiThreadedCodeIdentifierUtils.java
+++ b/spotbugs/src/main/java/edu/umd/cs/findbugs/util/MultiThreadedCodeIdentifierUtils.java
@@ -57,9 +57,6 @@ public class MultiThreadedCodeIdentifierUtils {
 
     public static boolean isPartOfMultiThreadedCode(ClassContext classContext) {
         JavaClass javaClass = classContext.getJavaClass();
-        if (notThreadSafe(javaClass.getAnnotationEntries())) {
-            return false;
-        }
         if (Subtypes2.instanceOf(javaClass, JAVA_LANG_RUNNABLE) ||
                 Stream.of(javaClass.getFields()).anyMatch(MultiThreadedCodeIdentifierUtils::isFieldIndicatingMultiThreadedContainer)) {
             return true;
@@ -155,8 +152,13 @@ public class MultiThreadedCodeIdentifierUtils {
         return false;
     }
 
-    private static boolean notThreadSafe(AnnotationEntry[] annotationEntries) {
-        for (AnnotationEntry annotationEntry : annotationEntries) {
+    /**
+     * @return <code>true</code> if the class is explicitely annotated with <code>NotThreadSafe</code> to document that it is not thread sage
+     */
+    public static boolean isNotThreadSafe(ClassContext classContext) {
+        JavaClass javaClass = classContext.getJavaClass();
+
+        for (AnnotationEntry annotationEntry : javaClass.getAnnotationEntries()) {
             if (annotationEntry.getAnnotationType().endsWith("/NotThreadSafe;")) {
                 return true;
             }

--- a/spotbugs/src/main/java/edu/umd/cs/findbugs/util/MultiThreadedCodeIdentifierUtils.java
+++ b/spotbugs/src/main/java/edu/umd/cs/findbugs/util/MultiThreadedCodeIdentifierUtils.java
@@ -153,7 +153,7 @@ public class MultiThreadedCodeIdentifierUtils {
     }
 
     /**
-     * @return <code>true</code> if the class is explicitely annotated with <code>NotThreadSafe</code> to document that it is not thread sage
+     * @return <code>true</code> if the class is explicitly annotated with <code>NotThreadSafe</code> to document that it is not thread safe
      */
     public static boolean isNotThreadSafe(ClassContext classContext) {
         JavaClass javaClass = classContext.getJavaClass();

--- a/spotbugs/src/main/java/edu/umd/cs/findbugs/util/MultiThreadedCodeIdentifierUtils.java
+++ b/spotbugs/src/main/java/edu/umd/cs/findbugs/util/MultiThreadedCodeIdentifierUtils.java
@@ -18,19 +18,19 @@
 
 package edu.umd.cs.findbugs.util;
 
-import edu.umd.cs.findbugs.ba.AnalysisContext;
-import edu.umd.cs.findbugs.ba.CFG;
-import edu.umd.cs.findbugs.ba.ClassContext;
-
 import java.util.Collection;
 import java.util.stream.Stream;
 
+import edu.umd.cs.findbugs.ba.AnalysisContext;
+import edu.umd.cs.findbugs.ba.CFG;
+import edu.umd.cs.findbugs.ba.ClassContext;
 import edu.umd.cs.findbugs.ba.DataflowAnalysisException;
 import edu.umd.cs.findbugs.ba.Location;
 import edu.umd.cs.findbugs.ba.LockDataflow;
 import edu.umd.cs.findbugs.ba.LockSet;
 import edu.umd.cs.findbugs.ba.ch.Subtypes2;
 import edu.umd.cs.findbugs.internalAnnotations.DottedClassName;
+import org.apache.bcel.classfile.AnnotationEntry;
 import org.apache.bcel.classfile.Field;
 import org.apache.bcel.classfile.JavaClass;
 import org.apache.bcel.classfile.LocalVariableTable;
@@ -57,6 +57,9 @@ public class MultiThreadedCodeIdentifierUtils {
 
     public static boolean isPartOfMultiThreadedCode(ClassContext classContext) {
         JavaClass javaClass = classContext.getJavaClass();
+        if (notThreadSafe(javaClass.getAnnotationEntries())) {
+            return false;
+        }
         if (Subtypes2.instanceOf(javaClass, JAVA_LANG_RUNNABLE) ||
                 Stream.of(javaClass.getFields()).anyMatch(MultiThreadedCodeIdentifierUtils::isFieldIndicatingMultiThreadedContainer)) {
             return true;
@@ -148,6 +151,15 @@ public class MultiThreadedCodeIdentifierUtils {
             }
         } catch (DataflowAnalysisException e) {
             AnalysisContext.logError(String.format("Synchronization check caught an error when analyzing %s method.", currentMethod.getName()), e);
+        }
+        return false;
+    }
+
+    private static boolean notThreadSafe(AnnotationEntry[] annotationEntries) {
+        for (AnnotationEntry annotationEntry : annotationEntries) {
+            if (annotationEntry.getAnnotationType().endsWith("/NotThreadSafe;")) {
+                return true;
+            }
         }
         return false;
     }

--- a/spotbugsTestCases/src/fakeAnnotations/javax/annotation/concurrent/NotThreadSafe.java
+++ b/spotbugsTestCases/src/fakeAnnotations/javax/annotation/concurrent/NotThreadSafe.java
@@ -1,8 +1,13 @@
 package javax.annotation.concurrent;
 
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
 import java.lang.annotation.Retention;
 import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
 
-@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Retention(RetentionPolicy.CLASS)
+@Target(ElementType.TYPE)
 public @interface NotThreadSafe {
 }

--- a/spotbugsTestCases/src/fakeAnnotations/javax/annotation/concurrent/NotThreadSafe.java
+++ b/spotbugsTestCases/src/fakeAnnotations/javax/annotation/concurrent/NotThreadSafe.java
@@ -1,0 +1,8 @@
+package javax.annotation.concurrent;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+@Retention(RetentionPolicy.RUNTIME)
+public @interface NotThreadSafe {
+}

--- a/spotbugsTestCases/src/java/ghIssues/Issue3390NotThreadSafe.java
+++ b/spotbugsTestCases/src/java/ghIssues/Issue3390NotThreadSafe.java
@@ -1,0 +1,17 @@
+package ghIssues;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+@NotThreadSafe
+public class Issue3390NotThreadSafe implements Runnable {
+    private int counter;
+
+    public int getCount() {
+        return counter;
+    }
+
+    @Override
+    public void run() {
+        counter++;
+    }
+}

